Firebase.cpp

```diff
@@ -15,100 +15,241 @@
   }
-  if (statusCode != 200) {
-    _error.set(statusCode,
-	       "stream " + location + ": "
-	       + HTTPClient::errorToString(statusCode));
+  // Remove is only complete if returned status is OK(200).
+  if (!call.isOk()) {
+    return FirebaseRemoveResult::FromError(
+        "Remove " + path + " returned with status " + call.httpStatus());
   }
-  return *this;
+  return FirebaseRemoveResult::Ok();
 }
 
-String Firebase::makeURL(const String& path) {
-  String url;
-  if (path[0] != '/') {
-    url = "/";
-  }
-  url += path + ".json";
-  if (_auth.length() > 0) {
-    url += "?auth=" + _auth;
+FirebaseEventStream Firebase::stream(const String& path) {
+  return FirebaseEventStream(host_, auth_, path);
+}
+
+/* FirebaseCall */
+FirebaseCall::FirebaseCall(const String& host, const String& auth,
+                           const char* method, const String& path, const String& value,
+                           HTTPClient* http) : http_(http) {
+  const String url = makeUrl(path, auth);
+  http_->begin(host.c_str(), kFirebasePort, url.c_str(), true, kFirebaseFingerprint);
+  status_ =  http_->sendRequest(method, (uint8_t*)value.c_str(), value.length());
+  if (isError()) {
+    error_message_ = String(method) + " " + url + ": " + HTTPClient::errorToString(status_);
   }
-  return url;
 }
 
-String Firebase::sendRequest(const char* method, const String& path, const String& value) {
-  _error.reset();
-  String url = makeURL(path);
-  _http.begin(_host.c_str(), firebasePort, url.c_str(), true, firebaseFingerprint);
-  int statusCode = _http.sendRequest(method, (uint8_t*)value.c_str(), value.length());
-  if (statusCode < 0) {
-    _error.set(statusCode,
-	       String(method) + " " + url + ": "
-	       + HTTPClient::errorToString(statusCode));
-    return "";
+FirebaseCall::FirebaseCall(const String& host, const String& auth,
+                           const char* method, const String& path,
+                           HTTPClient* http) : FirebaseCall(host, auth, method, path, "", http) {
+}
+
+bool FirebaseCall::isOk() const {
+  return status_ == HTTP_CODE_OK;
+}
+
+bool FirebaseCall::isError() const {
+  return status_ < 0;
+}
+
+String FirebaseCall::errorMessage() const {
+  return error_message_;
+}
+
+String FirebaseCall::rawResponse() {
+  return http_->getString();
+}
+
+/* FirebaseEventStream */
+
+FirebaseEventStream::FirebaseEventStream(const String& host, const String& auth,
+                                         const String& path) {
+  const String url = makeUrl(path, auth);
+  http_.setReuse(true);
+  http_.begin(host.c_str(), kFirebasePort, url.c_str(), true,
+              kFirebaseFingerprint);
+  const char* headers[] = {"Location"};
+  http_.collectHeaders(headers, 1);
+  http_.addHeader("Accept", "text/event-stream");
+  status_ = http_.sendRequest("GET", (uint8_t*)NULL, 0);
+
+  String location;
+  // TODO(proppy): Add a max redirect check
+  while (status_ == HTTP_CODE_TEMPORARY_REDIRECT) {
```

Comment:
I wonder if we should do this in the stream() call so we can get some early failure, and keep the dedicated client only for the request after the last redirect.